src/plugins/assets.ts

```diff
@@ -0,0 +1,72 @@
+import type { Plugin } from 'vite'
+
+import type { PWAPluginContext } from '../context'
+import { DEV_RELOAD_PAGE_NAME } from '../constants'
+
+export function AssetsPlugin(ctx: PWAPluginContext) {
+  return <Plugin>{
+    name: 'vite-plugin-pwa:assets',
+    enforce: 'post',
+    transformIndexHtml: {
+      order: 'post',
+      async handler(html) {
+        return await transformIndexHtmlHandler(html)
+      },
+      enforce: 'post', // deprecated since Vite 4
+      async transform(html) { // deprecated since Vite 4
+        return await transformIndexHtmlHandler(html)
+      },
+    },
+    async handleHotUpdate({ file, server }) {
+      const assetsGenerator = await ctx.assets
+      if (await assetsGenerator?.checkHotUpdate(file)) {
+        server.ws.send({
+          type: 'custom',
+          event: DEV_RELOAD_PAGE_NAME,
+        })
+        return []
+      }
+    },
+    configureServer(server) {
+      server.middlewares.use(async (req, res, next) => {
+        const url = req.url
+        if (!url)
+          return next()
+
+        const assetsGenerator = await ctx.assets
```

Comment:
if url isn't ico, svg, png, webp request just `return next()` (don't await and avoid extra lookups for any js/css resource)